modules/functional_expansion_tools/src/functions/FunctionSeries.C

```diff
@@ -0,0 +1,269 @@
+      types.push_back(_x);
+    }
+    if (isParamValid("y"))
+    {
+      domains.push_back(FunctionalBasisInterface::_domain_options = "y");
+      types.push_back(_y);
+    }
+    if (isParamValid("z"))
+    {
+      domains.push_back(FunctionalBasisInterface::_domain_options = "z");
+      types.push_back(_z);
+    }
+    if (types.size() == 0)
+      mooseError("Must specify one of 'x', 'y', or 'z' for 'Cartesian' series!");
+    _series_type = libmesh_make_unique<Cartesian>(domains, _orders, types, name());
+  }
+  else if (_series_type_name == "CylindricalDuo")
+  {
+    /*
+     * CylindricalDuo represents a disc-axial expansion, where the disc is described by a single
+     * series, such as Zernike (as opposed to a series individually representing r and a second
+     * series independently representing theta. For CylindricalDuo series, the series are always
+     * stored in the axial, planar order, independent of which order the series appear in the input
+     * file. Therefore, the _orders and _physical_bounds vectors must always appear in axial, planar
+     * order. The first entry in _domains is interpreted as the axial direction, and the following
+     * two as the planar.
+     */
+    if (isParamValid("x"))
+    {
+      domains = {FunctionalBasisInterface::_domain_options = "x",
+                 FunctionalBasisInterface::_domain_options = "y",
+                 FunctionalBasisInterface::_domain_options = "z"};
+      types.push_back(_x);
+    }
+    if (isParamValid("y"))
+    {
+      domains = {FunctionalBasisInterface::_domain_options = "y",
+                 FunctionalBasisInterface::_domain_options = "x",
+                 FunctionalBasisInterface::_domain_options = "z"};
+      types.push_back(_y);
+    }
+    if (isParamValid("z"))
+    {
+      domains = {FunctionalBasisInterface::_domain_options = "z",
+                 FunctionalBasisInterface::_domain_options = "x",
+                 FunctionalBasisInterface::_domain_options = "y"};
+      types.push_back(_z);
+    }
+
+    if (types.size() == 0)
+      mooseError("Must specify one of 'x', 'y', or 'z' for 'CylindricalDuo' series!");
+
+    if (types.size() > 1)
+      mooseError("Cannot specify more than one of 'x', 'y', or 'z' for 'CylindricalDuo' series!");
+
+    types.push_back(_disc);
+    _series_type = libmesh_make_unique<CylindricalDuo>(domains, _orders, types, name());
+  }
+  else
+    mooseError("Unknown functional series type \"", _series_type_name, "\"");
+
+  // Set the physical bounds of each of the single series if defined
+  if (isParamValid("physical_bounds"))
+    _series_type->setPhysicalBounds(_physical_bounds);
+
+  // Resize the coefficient array as needed
+  enforceSize(false), resize(getNumberOfTerms(), 0.0), enforceSize(true);
+  setCharacteristics(_orders);
+}
+
+FunctionSeries &
+FunctionSeries::checkAndConvertFunction(Function & function, const std::string & name)
+{
+  if (!dynamic_cast<FunctionSeries *>(&function))
```

Comment:
Use `libmesh_dynamic_cast` instead.  It compiles to a `static_cast` in optimized mode.

Reply:
Do you mean `libmesh_cast_ptr`?

I looks like it already performs an error check and report internally, so that simplifies this function to a one-liner (albeit with a little less contextual information)

Reply:
Looking through the use cases, the whole idea is to report to the user if they provided a `Function` object that is not a `FunctionSeries` type. It has nothing to do with ensuring the casting I'm doing during development is safe.

In other words, I need the `dynamic_cast` in all operational modes.

However, when looking as this I realized that many of my reconstruction objects perform their own checks. I will push an update soon that universally uses this static function instead.